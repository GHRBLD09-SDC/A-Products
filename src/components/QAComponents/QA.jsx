import React, { Component } from 'react';
import Question from './Question.jsx';
const api = 'http://3.134.102.30/qa';

class QA extends Component {
    constructor(props) {
        super(props);
        this.state = {
            searchTerm: '',
            questions: [],
            activeQuestions: [],
            currentPage: 0,
            moreToLoad: true
        }
        this.handleChange = this.handleChange.bind(this);
        this.fetchQuestions = this.fetchQuestions.bind(this);
        this.applyFilter = this.applyFilter.bind(this);
    }

    fetchQuestions(page, cb = () => { }) { 
        fetch(`${api}/${this.props.productId}?count=4&page=${page + 1}`)
            .then((res) => {
                res.json().then((data) => {
                    this.setState({
                        questions: this.state.questions.concat(data.results),
                        currentPage: page + 1,
                        moreToLoad: data.results.length > 0
                    }, () => {
                        cb()
                    });
                })
            });
    }

    componentDidMount() {
        this.fetchQuestions(this.state.currentPage, () => {
            this.setState({ activeQuestions: this.state.questions }); //initialize active questions to all questions
        });
    }

    // TODO: need to re-fetch questions when props.productId changes
    // getSnapshotBeforeUpdate() {
    //     if(window.scrollY) {
    //         return window.scrollY;
    //     } else {
    //         return null;
    //     }
    // }
    // componentDidUpdate(prevProps, prevState, snapshot) {
    //     if(snapshot !== null) {
    //         console.log(`scrolling to ${snapshot}`);
    //         window.scrollTo(0, snapshot);
    //     }
    // }

    handleChange(e) {
        this.setState({ searchTerm: e.target.value }, () => {
            if (this.state.searchTerm.length > 2) {
                this.applyFilter()
            } else if (this.state.searchTerm.length === 0) {
                this.setState({ activeQuestions: this.state.questions });
            }
        });
    }

    applyFilter() {
        const searchTerm = this.state.searchTerm.length > 2 ? this.state.searchTerm : '';
        const filteredQs = [...this.state.questions].filter(question => {
            if (question.question_body.includes(searchTerm)) {
                return true;
            } else {
                return Object.values(question.answers).some(answer => answer.body.includes(searchTerm));
            }
        });
        this.setState({ activeQuestions: filteredQs });
    }

    refresh() {
        // console.log('setting state')
        this.setState({
            questions: [],
            activeQuestions: []
        }, () => {
            // console.log('rerendering...')
            const { currentPage } = this.state;
            let tempPage = 0;
            while (tempPage < currentPage) {
                this.fetchQuestions(tempPage, this.applyFilter); //using Promise.all here might be more performant...
                tempPage++;
            }
        });
    }

    render() {
        return (
            <div>
                <h3>Questions and Answers</h3>
                <form className="searchQuestions">
                    <input type="text"
                        className="q-search"
                        placeholder="HAVE A QUESTION? SEARCH FOR ANSWERS."
                        value={this.state.searchTerm}
                        onChange={this.handleChange}
                    />
                </form>
                {this.state.activeQuestions.map((question) => <Question
                    question={question}
                    key={question.question_id}
                    updateParent={this.refresh.bind(this)} />)}
                {this.state.moreToLoad && <button onClick={() => this.fetchQuestions(this.state.currentPage, this.applyFilter)}>MORE ANSWERERD QUESTIONS</button>}
                <button>ASK A QUESTION +</button>
            </div>
        )
    }
}

export default QA;