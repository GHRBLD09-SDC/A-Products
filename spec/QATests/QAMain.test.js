import React from 'react';
import QA from '../../src/components/QAComponents/QA';
import data from './SampleData';
import renderer from 'react-test-renderer';
import { render, fireEvent, wait } from '@testing-library/react';

const mockData = Promise.resolve(data.questions);
window.fetch = jest.fn(() => Promise.resolve({
    json: () => mockData
}));

describe('QA Component', () => {
    it('Renders the Q&A component correctly', () => {
        const QATest = renderer.create(<QA productId={5}></QA>);
        expect(QATest.toJSON()).toMatchSnapshot();
    });

    it('Fetches questions and renders them on the page', async () => {
        const QATest = render(<QA productId={5}/>);
        const { getAllByTestId, findByTestId } = QATest;
        await wait(() => {
            findByTestId('question');
        })
        const questions = getAllByTestId('question');
        expect(questions.length).toBe(4);
    });
});


