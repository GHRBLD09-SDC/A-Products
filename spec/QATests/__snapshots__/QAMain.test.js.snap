// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QA Component Renders the Q&A component correcrtly 1`] = `
<div
  data-widget="Q&A-5"
>
  <p
    className="text-main"
    data-selector="qa-title"
  >
    QUESTIONS & ANSWERS
  </p>
  <div
    className="questions-container"
  >
    <p
      className="text-reg"
      data-selector="no-questions"
    >
      Be the first to ask a question about this product!
    </p>
  </div>
  <button
    data-selector="new-question"
    onClick={[Function]}
  >
    ASK A QUESTION +
  </button>
</div>
`;

exports[`QA Component Renders the Q&A component correctly 1`] = `
<div
  data-widget="Q&A-5"
>
  <p
    className="text-main"
    data-selector="qa-title"
  >
    QUESTIONS & ANSWERS
  </p>
  <div
    className="questions-container"
  >
    <p
      className="text-reg"
      data-selector="no-questions"
    >
      Be the first to ask a question about this product!
    </p>
  </div>
  <button
    data-selector="new-question"
    onClick={[Function]}
  >
    ASK A QUESTION +
  </button>
</div>
`;

exports[`QA Snapshot Renders the Q&A component correcrtly 1`] = `
<div
  data-widget="Q&A-5"
>
  <p
    className="text-main"
    data-selector="qa-title"
  >
    QUESTIONS & ANSWERS
  </p>
  <div
    className="questions-container"
  >
    <p
      className="text-reg"
      data-selector="no-questions"
    >
      Be the first to ask a question about this product!
    </p>
  </div>
  <button
    data-selector="new-question"
    onClick={[Function]}
  >
    ASK A QUESTION +
  </button>
</div>
`;

exports[`Renders the Q&A component correcrtly 1`] = `
<div
  data-widget="Q&A-5"
>
  <p
    className="text-main"
    data-selector="qa-title"
  >
    QUESTIONS & ANSWERS
  </p>
  <div
    className="questions-container"
  >
    <p
      className="text-reg"
      data-selector="no-questions"
    >
      Be the first to ask a question about this product!
    </p>
  </div>
  <button
    data-selector="new-question"
    onClick={[Function]}
  >
    ASK A QUESTION +
  </button>
</div>
`;
